### ISSUE TITLE
make check and make install fail on macOS using (mini)conda virtualenv and Python 3.4/3.5

### ISSUE THREAD

**Issue by marvel64** [TGAC/KAT#78]
Hi,

Compiling version 2.3.1 on macOS using a conda (miniconda) virtual environment fails due to
problems loading `libpython3.5m.dylib`.

All steps were executed within the activated python 3.5 virtual environment using `./configure` without additional options. The dependencies were installed via `homebrew` or `conda`. I tried
Python 3.4 also to no avail.

Details follow below. Any help or hint in the right direction would be highly appreciated.

Thank you!
Marc

macOS 10.12.3
Xcode command line tools version: 8.2.0.0.1.1480973914

The `./configure` step seems to be ok.

```
...skipped lines
checking for python3... /Users/Marc/miniconda/envs/py3.5/bin/python3
configure: Found python version: 3.5.2
checking for a version of Python >= '3.1.0'... yes
checking for a version of Python >= '3.1'... yes
configure: Ensuring environment variable 'PYTHON_VERSION' set to: 3
checking for the distutils Python package... yes
checking for Python include path... -I/Users/Marc/miniconda/envs/py3.5/include/python3.5m
checking for Python library path... -L/Users/Marc/miniconda/envs/py3.5/lib -lpython3.5m
checking for Python site-packages path... /Users/Marc/miniconda/envs/py3.5/lib/python3.5/site-packages
checking python extra libraries... -ldl  -framework CoreFoundation
checking python extra linking flags... -Wl,-stack_size,1000000  -framework CoreFoundation
checking consistency of all components of python development environment... yes
checking python3 module: numpy... yes
checking python3 module: matplotlib... yes
checking python3 module: scipy... yes
checking for sphinx-build... yes
... skipped lines
configure: Using python plotting
configure: Building documentation with sphinx
configure: Boost will be statically linked to your executable
```

As is `make`.

```
...skipped lines
Making all in tests
make[2]: Nothing to be done for `all'.
Making all in doc
sphinx-build -b html -d build/doctrees   -c source ../doc/source build/html
Running Sphinx v1.5.1
making output directory...
loading pickled environment... not yet created
WARNING: latex_use_parts is deprecated. Use latex_toplevel_sectioning instead.
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 6 source files that are out of date
updating environment: 6 added, 0 changed, 0 removed
reading sources... [100%] walkthrough
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] walkthrough
generating indices... genindex
writing additional pages... search
copying images... [100%] images/pe_v_mp_before_shared.png
copying static files... WARNING: html_static_path entry '/usr/local/kat-2.3.1/doc/source/_static' does not exist
done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 2 warnings.

Build finished. The HTML pages are in build/html.
sphinx-build -b man -d build/doctrees   -c source ../doc/source build/man
Running Sphinx v1.5.1
making output directory...
loading pickled environment... done
WARNING: latex_use_parts is deprecated. Use latex_toplevel_sectioning instead.
building [mo]: targets for 0 po files that are out of date
building [man]: all manpages
updating environment: 0 added, 0 changed, 0 removed
looking for now-outdated files... none found
writing... kat.1 { installation using kmer walkthrough faq }
build succeeded, 1 warning.

Build finished. The manual pages are in build/man.
make[2]: Nothing to be done for `all-am'.
```

But `make check` fails:

```
PASS: tests/generate_sequence.sh
PASS: tests/parallel_hashing.sh
PASS: tests/merge.sh
SKIP: tests/big.sh
PASS: tests/subset_hashing.sh
PASS: tests/multi_file.sh
PASS: tests/large_key.sh
PASS: unit_tests/unit_tests.sh
/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-recursive
Making all in .
make[8]: Nothing to be done for `all-am'.
============================================================================
Testsuite summary for kat_jellyfish 2.2.0
============================================================================
# TOTAL: 8
# PASS:  7
# SKIP:  1
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
Making check in deps/seqan-library-2.0.0
make[1]: Nothing to be done for `check'.
Making check in lib
make[1]: Nothing to be done for `check'.
Making check in src
make[1]: Nothing to be done for `check'.
Making check in tests
/Applications/Xcode.app/Contents/Developer/usr/bin/make  libgtest.la libgtest_main.la check_unit_tests
  CXX      gtest/src/libgtest_la-gtest-all.lo
  CXXLD    libgtest.la
  CXX      gtest/src/libgtest_main_la-gtest_main.lo
  CXXLD    libgtest_main.la
  CXX      check_jellyfish.o
  CXX      check_spectra_helper.o
  CXX      check_compcounters.o
  CXX      check_main.o
  CXXLD    check_unit_tests
clang: warning: argument unused during compilation: '-pthread'
clang: warning: argument unused during compilation: '-pthread'
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
../build-aux/test-driver: line 107: 36903 Abort trap: 6           "$@" > $log_file 2>&1
FAIL: check_unit_tests
FAIL: test_hist.sh
FAIL: test_gcp.sh
FAIL: test_sect.sh
FAIL: test_comp.sh
make[4]: Nothing to be done for `all'.
============================================================================
Testsuite summary for kat 2.3.1
============================================================================
# TOTAL: 5
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  5
# XPASS: 0
# ERROR: 0
============================================================================
See tests/test-suite.log
Please report to https://github.com/TGAC/KAT/issues
============================================================================
make[3]: *** [test-suite.log] Error 1
make[2]: *** [check-TESTS] Error 2
make[1]: *** [check-am] Error 2
make: *** [check-recursive] Error 1
```

All errors in `test-suite.log` are related to `libpython3.5m.dylib`, only one entry included below.

```
FAIL: check_unit_tests
======================

dyld: Library not loaded: @rpath/libpython3.5m.dylib
  Referenced from: /usr/local/kat-2.3.1/tests/./check_unit_tests
  Reason: image not found
```

**Comment by maplesond** [comment]
To me this seems like the python library, while available during compilation time, library isnt available at runtime to be loaded by the test programs.  I'm not that familiar with macs but on unix systems we would normally set a  variable LD_LIBRARY_PATH to do this.  Could you double check this and let me know if there is still a problem?  Best, Dan

**Comment by marvel64** [comment]
Hi,

Thank you for getting back!

Setting LD_LIBRARY_PATH (.zshrc) worked on a different machine:

```
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/Users/Marc/miniconda/envs/py3.5/lib/
```

On the first machine the `make check` and `make install` steps still failed but I am currently unsure why. There, the only solution was installing `kat` via homebrew.

However, on the second machine I get this error for all commands I tested (except filter/plot) after the kmer analysis: 

```
Creating plot ...Could not find platform independent libraries <prefix>
Could not find platform dependent libraries <exec_prefix>
Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
Fatal Python error: Py_Initialize: unable to load the file system codec
ImportError: No module named 'encodings’
```

Any idea on this?

Thank you very much!
Marc

**Comment by maplesond** [comment]
Hi Marc,

I've not seen this issues before.  I don't have a mac on my desktop, and normally we wouldn't heavily target Mac as KAT can have quite high memory requirements, which typically are only satisfied by servers or other HPC environments.  Having said that, I'm keen to try and get to the bottom of this one.

Out of interest do you get the same problems with the homebrew version of KAT? Or is this just with your own version compiled from source?

You can test to see what runtime dependencies an executable has by typing "ldd <executable>".  If the environment is not configured correctly then you can see if any dependencies are not satisfied.  Hopefully this gives you an idea if this is a environment configuration issue or not.

By the way, I've just released a new version of KAT v2.3.2.  I'm in the process of updating homebrew so hopefully you can get the latest version shortly that way.

Best,
Dan

**Comment by marvel64** [comment]
Hi Dan,

`kat` installed through homebrew works on the one machine I tested it. I was a little reluctant to do it this way as it will install python3, numpy, spicy, and matplotlib as well, and I did not want it to collide with my conda installation.
But for the few times I ran `kat` I had no issues.

I will try if I can resolve the issue with the version compiled from source.

Thanks for updating the homebrew version and for your help!
Marc
